arch/x86/kvm/vmx/pkvm/hyp/iommu_spgt.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0
+ * Copyright(c) 2022 Intel Corporation.
+ * Copyright(c) 2023 Semihalf.
+ */
+
+#include <linux/hashtable.h>
+#include <pkvm_spinlock.h>
+#include <pkvm.h>
+#include <gfp.h>
+#include "pkvm_hyp.h"
+#include "iommu_spgt.h"
+#include "ept.h"
+#include "bug.h"
+
+static DEFINE_HASHTABLE(iommu_spgt_hasht, 8);
+static DECLARE_BITMAP(iommu_spgt_bitmap, PKVM_MAX_PDEV_NUM);
+static struct pkvm_iommu_spgt pkvm_iommu_spgt[PKVM_MAX_PDEV_NUM];
+static pkvm_spinlock_t iommu_spgt_lock = { __ARCH_PKVM_SPINLOCK_UNLOCKED };
+
+struct pkvm_pgtable *pkvm_get_iommu_spgt(unsigned long root_gpa, bool coherency)
+{
```

Comment:
> pkvm_host_get_iommu_spgt() as it's only for host VM?

I'm fine with it, OTOH maybe this get/put API does not need to be aware of the fact that it's used only for the host?

> BTW, this patch could be separated into two patches, one for memory reservation, another one for adding host shadow iommu page table APIs.

Agree.

Reply:
> > pkvm_host_get_iommu_spgt() as it's only for host VM?
> 
> I'm fine with it, OTOH maybe this get/put API does not need to be aware of the fact that it's used only for the host?
> 

Yes, I agree. If so, could we combine it with pgstate_pgt? The pgstate pgt should also be able to allocate from this API.

Reply:
Hmm, interesting idea. But it seems we cannot just adapt this API for pgstate_pgt as is, since it assumes 1:1 correspondence between vIOMMU page table (identified by its root address) and shadow IOMMU page table, while pgstate_pgt is per VM, and as I understand, each VM needs its own pgstate_pgt for storing its own page states, even if different VMs share the same vIOMMU page table.

So now I'm thinking, maybe it's indeed better to name it pkvm_host_get_iommu_spgt() as you suggest, since this API as is seems suitable for the host only.

Reply:
After you know a ptdev belong to a guest VM, and know ptdev->pgt is pgstate pgt, then you can directly return pgstate pgt for this API - which means you will not touch your spgt logic.

Is it ok?

Reply:
Are you suggesting to rework the way how we set ptdev->pgt to pgstate_pgt for an attached ptdev: instead of doing it in pkvm_attach_ptdev(), do it through this generalized get_iommu_spgt API which will decide whether the given ptdev shall use pgstate_pgt (if attached) or a specially allocated iommu spgt (if not attached)? If so, get_iommu_spgt would need access to the pkvm_shadow_vm structure (to get &vm->pgstate_pgt) which is currently not easily available through pkvm_ptdev structure.

Overall, this sounds like a significant reorganization of the existing code. And maybe in the end it would make the code cleaner, or maybe quite the opposite, it would make it just more confusing.

I was actually intending just to add support for legacy IOMMU shadow tables, without making drastic changes in the existing code for pgstate_pgt.

Reply:
Yes, I agree, currently we do not need to do big refactor for the code.

Reply:
Renamed to `pkvm_get_host_iommu_spgt()` and `pkvm_put_host_iommu_spgt()`.